core/src/main/java/io/aiven/klaw/service/JwtTokenUtilService.java

```diff
@@ -0,0 +1,90 @@
+package io.aiven.klaw.service;
+
+import io.aiven.klaw.error.KlawNotAuthorizedException;
+import io.jsonwebtoken.Claims;
+import io.jsonwebtoken.Jws;
+import io.jsonwebtoken.Jwts;
+import io.jsonwebtoken.SignatureAlgorithm;
+import java.security.Key;
+import java.util.Date;
+import java.util.List;
+import java.util.function.Function;
+import javax.crypto.spec.SecretKeySpec;
+import org.apache.tomcat.util.codec.binary.Base64;
+import org.springframework.beans.factory.InitializingBean;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.stereotype.Service;
+
+@Service
+@ConditionalOnProperty(prefix = "klaw.core.ha", name = "enable", matchIfMissing = false)
+public class JwtTokenUtilService implements InitializingBean {
+  @Value("${klaw.core.app2app.base64.secret:#{''}}")
+  private String App2AppSecret;
+
+  public static final String BEARER = "Bearer ";
+  @Autowired UserDetailsService userDetailsService;
+  private static byte[] decodedSecret;
+
+  // retrieve username from jwt token
+  public String getUsernameFromToken(String token) {
+    return getClaimFromToken(token, Claims::getSubject);
+  }
+
+  // retrieve expiration date from jwt token
+  private Date getExpirationDateFromToken(String token) {
+    return getClaimFromToken(token, Claims::getExpiration);
+  }
+
+  private <T> T getClaimFromToken(String token, Function<Claims, T> claimsResolver) {
+    final Claims claims = getAllClaimsFromToken(token);
+    return claimsResolver.apply(claims);
+  }
+
+  // for retrieving any information from token we will need the secret key
+  public Claims getAllClaimsFromToken(String token) {
+    Key hmacKey = new SecretKeySpec(decodedSecret, SignatureAlgorithm.HS256.getJcaName());
+    Jws<Claims> jwt = Jwts.parserBuilder().setSigningKey(hmacKey).build().parseClaimsJws(token);
+    return jwt.getBody();
+  }
+
+  // check if the token has expired
+  private Boolean isTokenExpired(String token) {
+    final Date expiration = getExpirationDateFromToken(token);
+    return expiration.before(new Date());
+  }
+
+  // validate token
+  public Boolean validateToken(String token) {
+    return !isTokenExpired(token);
+  }
+
+  public void validateRole(String token, String... expectedRole) throws KlawNotAuthorizedException {
+
+    token = token.substring(BEARER.length()).trim();
+    Claims claims = getAllClaimsFromToken(token);
+    UserDetails user = userDetailsService.loadUserByUsername(getUsernameFromToken(token));
+    if (user == null
+        && !validateToken(token)
+        && !claims.get("Roles", List.class).containsAll(List.of(expectedRole))) {
```

Comment:
- raw use of List class
- do we need 'Roles' here ? didn't understand

Reply:
Roles here makes sure that the correct Roles have been assigned to the user, so another user might come in with a valid token but might not have the role CACHE_ADMIN and will receive an unauthorized response.
So its adding the role at the API level to prevent unauthorized access.